File: Quiz_3/Nakul_500121882/Quiz_3_Debugging_Ques1.c
// Buggy version
int binarySearch(int nums[], int n, int target)
{
    int left = 0, right = n;  // ❌ Bug: right should be n - 1

    while (left < right)
    {
        int mid = (left + right) / 2;
        if (nums[mid] == target)
            return mid;
        else if (nums[mid] < target)
            left = mid;  // ❌ Bug: can cause infinite loop
        else
            right = mid - 1;
    }
    return -1;
}

// 1. Bugs
// - `right = n` is invalid since C arrays are 0-indexed. Use `right = n - 1`.
// - `left = mid` is incorrect when left == mid → leads to infinite loop.

// 2. Consider int nums[] = {1, 3, 5, 7}; target = 7;
// Expected output = 3 (index of 7)
// Buggy version may cause an infinite loop or return wrong result.

// 3. The above function may return incorrect result or loop forever.

// 4. Corrected version
int binarySearch(int nums[], int n, int target)
{
    int left = 0, right = n - 1;  // ✅ Fix: inclusive right

    while (left <= right)  // ✅ Fix: include right
    {
        int mid = (left + right) / 2;
        if (nums[mid] == target)
            return mid;
        else if (nums[mid] < target)
            left = mid + 1;  // ✅ Fix: move past mid
        else
            right = mid - 1;
    }
    return -1;
}

File: Quiz_3/Angel_500119630/Quiz_3_Debugging_Ques1.py

# Buggy version
def binary_search(nums, target):
    left, right = 0, len(nums)  # Bug: right should be len(nums) - 1

    while left < right:
        mid = (left + right) // 2
        if nums[mid] == target:
            return mid
        elif nums[mid] < target:
            left = mid      # Bug: should be mid + 1
        else:
            right = mid - 1
    return -1

# 1. Bugs
# - right = len(nums) causes out-of-bound logic since last valid index is len(nums) - 1
# - left = mid causes infinite loop when left == mid

# 2. Consider nums = [1, 3, 5, 7], target = 7
# Expected output = 3 (index of 7)
# Buggy version may return -1 or loop forever

# 3. The above function may return incorrect index or never terminate

# 4. Corrected version
def binary_search(nums, target):
    left, right = 0, len(nums) - 1  # Fix: set right to last valid index

    while left <= right:  # Include both ends
        mid = (left + right) // 2
        if nums[mid] == target:
            return mid
        elif nums[mid] < target:
            left = mid + 1  # Move past mid
        else:
            right = mid - 1
    return -1
